thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java

```diff
@@ -0,0 +1,320 @@
+   */
+  private boolean isMissingData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info) {
+    return datasetLastRefreshTime < info.getStart();
+  }
+
+  /**
+   * We say the data is partial if we do not have all the data points in the sla detection window.
+   * Or more specifically if we have at least 1 data-point missing in the sla detection window.
+   *
+   * For example:
+   * Assume that the data is delayed and our current sla detection window is [1st Feb to 3rd Feb). During this scan,
+   * let's say data for 1st Feb arrives. Now, we have a situation where partial data is present. In other words, in the
+   * current sla detection window [1st to 3rd Feb) we have data for 1st Feb but data for 2nd Feb is missing. This is the
+   * partial data scenario.
+   */
+  private boolean isPartialData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info, DatasetConfigDTO datasetConfig) {
+    long granularity = datasetConfig.bucketTimeGranularity().toMillis();
+    return (info.getEnd() - datasetLastRefreshTime) * 1.0 / granularity > 1;
+  }
+
+  /**
+   * Validates if the data is delayed or not based on the user specified SLA configuration
+   */
+  private boolean hasMissedSLA(long datasetLastRefreshTime, Map<String, Object> slaProps, long slaDetectionEndTime) {
+    // fetch the user configured SLA, otherwise default 3_DAYS.
+    long delay = TimeGranularity.fromString(MapUtils.getString(slaProps,  "sla", DEFAULT_DATA_SLA))
+        .toPeriod().toStandardDuration().getMillis();
+
+    return (slaDetectionEndTime - datasetLastRefreshTime) >= delay;
+  }
+
+  /**
+   * Align and round off start time to the upper boundary of the granularity
+   */
+  private static long alignToUpperBoundary(long start, DatasetConfigDTO datasetConfig) {
+    Period granularityPeriod = datasetConfig.bucketTimeGranularity().toPeriod();
+    DateTimeZone timezone = DateTimeZone.forID(datasetConfig.getTimezone());
+    DateTime startTime = new DateTime(start - 1, timezone).plus(granularityPeriod);
+    return startTime.getMillis() / granularityPeriod.toStandardDuration().getMillis() * granularityPeriod.toStandardDuration().getMillis();
+  }
+
+  /**
+   * Merges one DATA_MISSING anomaly with remaining existing anomalies.
+   */
+  private void mergeSLAAnomalies(MergedAnomalyResultDTO anomaly, List<MergedAnomalyResultDTO> existingAnomalies) {
+    // Extract the parent SLA anomaly. We can have only 1 parent DATA_MISSING anomaly in a window.
+    existingAnomalies.removeIf(MergedAnomalyResultBean::isChild);
+    MergedAnomalyResultDTO existingParentSLAAnomaly = existingAnomalies.get(0);
+
+    if (isDuplicateSLAAnomaly(existingParentSLAAnomaly, anomaly)) {
+      // Ensure anomalies are not duplicated. Ignore and return.
+      // Example: daily data with hourly cron should generate only 1 sla alert if data is missing
+      return;
+    }
+    existingParentSLAAnomaly.setChild(true);
+    anomaly.setChild(false);
```

Comment:
Why this is called "merge"? It is only dedup?

Reply:
Discussed offline. When 2 sla anomalies overlap (not duplicate) we merge them into a parent child relationship.